### PR TITLE
SNOW-25773: Query Pushdowns

### PR DESCRIPTION
**Pull request by etduwx**

### CODE REVIEW

File: src/it/scala/net/snowflake/spark/snowflake/benchmarks/TCPHSuite.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import net.snowflake.spark.snowflake.{Conversions, SnowflakeInputFormat}
+import net.snowflake.spark.snowflake.Utils._
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.types._
+
+import scala.collection.mutable
+
+class TCPHSuite extends PerformanceSuite {
```
Comment (MarcinZukowski) [inline src/it/scala/net/snowflake/spark/snowflake/benchmarks/TCPHSuite.scala]: It's TPCH :)

File: src/it/scala/net/snowflake/spark/snowflake/SimpleNewPushdownIntegrationSuite.scala [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+      super.afterAll()
+      SnowflakeConnectorUtils.disablePushdownSession(sqlContext.sparkSession)
+    }
+  }
+
+  /**
+    * Verify that the pushdown was done by looking at the generated SQL,
+    * and check the results are as expected
+    */
+  def testPushdown(reference: String,
```
Comment (MarcinZukowski) [inline src/it/scala/net/snowflake/spark/snowflake/SimpleNewPushdownIntegrationSuite.scala]: You have another, almost identical testPushdown() in another file, maybe merge them

File: src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+      }
+      case e: Exception => throw e
+    }
+
+    if (runTests)
+      verifyParams()
+  }
+
+  protected final def verifyParams(): Unit = {
```
Comment (MarcinZukowski) [inline src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala]: general comment: some more comments would be nice...

File: src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                                name: String = "unnamed"): Unit = {
+
+    val failedMessage = s"""Failed: $name. No result generated."""
+
+    // Skip if invalid configs
+    if (!runTests) return
+
+    var result: String = ""
+
+    if (runOption == "both") {
```
Comment (MarcinZukowski) [inline src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala]: Almost always a good idea to introduce constants for repeated strings.

File: src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+      result = runWithSnowflake(query, name, true).getOrElse(failedMessage)
+      outputResult(s"""${name.toUpperCase}, with full pushdowns: """ + result)
+
+      result = runWithSnowflake(query, name, false).getOrElse(failedMessage)
+      outputResult(
+        s"""${name.toUpperCase}, with only filter/proj pushdowns: """ + result)
+    } else if (runOption == "full-only") {
```
Comment (MarcinZukowski) [inline src/it/scala/net/snowflake/spark/snowflake/benchmarks/PerformanceSuite.scala]: I'd remove the branch above, and simply make this one 
  if  (runOption == "full-only" or runOption == "both")
and similar for the else below.

File: src/main/scala/net/snowflake/spark/snowflake/Conversions.scala [hunk 6]
```diff
@@ -22,20 +22,27 @@ import java.text._
...
   // For TZ and LTZ, Snowflake serializes with timezone
   // Note - we use a pattern with timezone in the beginning, to make sure
   // parsing with PATTERN_NTZ fails for PATTERN_TZLTZ strings.
   // Note - for JDK 1.6, we use Z ipo XX for SimpleDateFormat
   private val PATTERN_TZLTZ =
-      if (System.getProperty("java.version").startsWith("1.6."))
-          "Z yyyy-MM-dd HH:mm:ss.SSS" else "XX yyyy-MM-dd HH:mm:ss.SSS"
+    if (System.getProperty("java.version").startsWith("1.6."))
+      "Z yyyy-MM-dd HH:mm:ss.SSS"
+    else "XX yyyy-MM-dd HH:mm:ss.SSS"
```
Comment (MarcinZukowski) [inline src/main/scala/net/snowflake/spark/snowflake/Conversions.scala]: nit: nl after else for formatting consistency

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 7]
```diff
@@ -69,7 +69,9 @@ private[snowflake] class JDBCWrapper {
     * @throws SQLException if the table contains an unsupported type.
     */
   def resolveTable(conn: Connection, table: String): StructType = {
-    val rs = executeQueryInterruptibly(conn, s"SELECT * FROM $table WHERE 1=0")
+    Math.abs(Random.nextLong()).toString
+    val randStr = Math.abs(Random.nextLong()).toString
+    val rs = executeQueryInterruptibly(conn, s"SELECT * FROM ($table) as temp_$randStr WHERE 1=0")
```
Comment (MarcinZukowski) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala]: Not sure, why randstr is needed here?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 8]
```diff
@@ -300,7 +302,13 @@ private[snowflake] class JDBCWrapper {
       case java.sql.Types.CLOB => StringType
       case java.sql.Types.DATALINK => null
       case java.sql.Types.DATE => DateType
-      case java.sql.Types.DECIMAL if precision != 0 || scale != 0 => DecimalType(precision, scale)
+      case java.sql.Types.DECIMAL if precision != 0 || scale != 0 => {
+        if (precision > DecimalType.MAX_PRECISION) {
+          DecimalType(DecimalType.MAX_PRECISION, scale + (precision - DecimalType.MAX_SCALE))
```
Comment (MarcinZukowski) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala]: Not sure why we'd increase scale here. I understand we might limit precision, but doesn't it mean scale should still stay the same?

File: src/main/scala/net/snowflake/spark/snowflake/pushdowns/querygeneration/AggregationExpression.scala [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package net.snowflake.spark.snowflake.pushdowns.querygeneration
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate._
+
+/**
+  * Extractor for aggregate-style expressions.
```
Comment (MarcinZukowski) [inline src/main/scala/net/snowflake/spark/snowflake/pushdowns/querygeneration/AggregationExpression.scala]: nit: you're inconsistent with your comments, sometimes you start in the first line, sometimes in the second

File: src/main/scala/net/snowflake/spark/snowflake/pushdowns/querygeneration/QueryHelper.scala [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    p.map(e => convertExpression(e, colSet)).mkString(", ")
+  }
+
+  val output: Seq[Attribute] = {
+    processedProjections.getOrElse {
+
+      if (children.isEmpty) {
+        outputAttributes.getOrElse(throw new SnowflakePushdownException(
+          "Query output attributes must not be empty when it has no children."))
+
```
Comment (MarcinZukowski) [inline src/main/scala/net/snowflake/spark/snowflake/pushdowns/querygeneration/QueryHelper.scala]: nit: remove nl

**Comment by etduwx** [comment]
@MarcinZukowski Thanks a lot for the review and helpful comments. As you can see, I'm actually still making commits and modifying this branch, so will hold off on merging until I'm all done, but will also address your feedback before merge as well.

**Pull request merged**